core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,341 @@
+          inputStream = mUnderFileInputStreamCache.getIfPresent(id);
+          if (inputStream != null) {
+            nextId = id;
+            LOG.debug("Reused the under file input stream resource of {}", nextId);
+            // for the cached ufs instream, seek to the requested position
+            inputStream.seek(openOptions.getOffset());
+            break;
+          }
+        }
+      }
+      // no cached input stream is available, open a new one
+      if (nextId == UNAVAILABLE_RESOURCE_ID) {
+        nextId = IdUtils.getRandomNonNegativeLong();
+        final long newId = nextId;
+        try {
+          inputStream = mUnderFileInputStreamCache.get(nextId, () -> {
+            SeekableUnderFileInputStream ufsStream = (SeekableUnderFileInputStream) ufs.open(path,
+                OpenOptions.defaults().setOffset(openOptions.getOffset()));
+            LOG.debug("Created the under file input stream resource of {}", newId);
+            return new CachedSeekableInputStream(ufsStream, newId, fileId, path);
+          });
+        } catch (ExecutionException e) {
+          LOG.warn("Failed to retrieve the cached UFS instream");
+          // fall back to a ufs creation.
+          return ufs.open(path, OpenOptions.defaults().setOffset(openOptions.getOffset()));
+        }
+      }
+
+      // mark the input stream id as acquired
+      resources.acquire(nextId);
+      return inputStream;
+    }
+  }
+
+  private boolean isCachingEnabled() {
+    return Configuration.getBoolean(PropertyKey.WORKER_UFS_INSTREAM_CACHE_ENABLE);
+  }
+
+  /**
+   * The metadata of the input streams associated with an under storage file that tracks which input
+   * streams are in-use or available. Each input stream is identified by a unique id.
+   */
+  @ThreadSafe
+  private static class UfsInputStreamIds {
+    private final Set<Long> mInUse;
+    private final Set<Long> mAvailable;
+
+    /**
+     * Creates a new {@link UfsInputStreamIds}.
+     */
+    UfsInputStreamIds() {
+      mInUse = new HashSet<>();
+      mAvailable = new HashSet<>();
+    }
+
+    /**
+     * @return a view of the available input stream ids
+     */
+    Set<Long> availableIds() {
+      return Collections.unmodifiableSet(mAvailable);
+    }
+
+    /**
+     * Marks an input stream as acquired.
+     *
+     * @param id the id of the input stream
+     */
+    public synchronized void acquire(long id) {
+      Preconditions.checkArgument(!mInUse.contains(id), id + " is already in use");
+      mAvailable.remove(id);
+      mInUse.add(id);
+    }
+
+    /**
+     * @return if there is any outstanding input streams of the file
+     */
+    public synchronized boolean isEmpty() {
+      return mInUse.isEmpty() && mAvailable.isEmpty();
+    }
+
+    /**
+     * Marks an input stream as not in use.
+     * @param id the id of the input stream
+     */
+    public synchronized void removeInUse(long id) {
+      mInUse.remove(id);
```

Comment:
what does it mean by not in use and not available yet?

Reply:
It should not be possible. You can think of this class as an index to the resources in the cache. For a resource in the cache, it's either in use or available.

Reply:
@yupeng9 then why we expose `removeInUse` which may lead to this state and up to the user to ensure the atomicity?

Reply:
No, this is not exposed as `UfsInputStreamIds` is private. I'll drop the `public` modifier.